File: go/mergeSort.go
// マージソート
package main

import (
	"fmt"
	"math"
)

func main() {
    arr := []int{148, 651, 124, 638, 567, 435, 185, 413, 841, 35};
    fmt.Println(margeSort(arr));
}
func margeSort(arr []int) []int {
    var m func(int, int) []int;
    m = func(head int, tail int) []int {
        if head < tail {
            mid := int(math.Floor(float64(head + tail) / 2));
            m(head, mid);
            m(mid + 1, tail);

            l := make([]int, len(arr[head:mid+1]), (cap(arr[head:mid+1])+1))
            r := make([]int, len(arr[mid+1:tail+1]), (cap(arr[mid+1:tail+1])+1))
            copy(l, arr[head:mid+1]);
            copy(r, arr[mid+1:tail+1]);

            i := head;
            for len(l) > 0 && len(r) > 0 {
                if l[0] <= r[0] {
                    arr[i] = l[0];
                    l = l[1:];
                } else {
                    arr[i] = r[0];
                    r = r[1:];
                }
                i++;
            }
            
            for len(l) > 0 {
                arr[i] = l[0];
                l = l[1:];
                i++;
            }
        }
        return arr;
    }
    return m(0, len(arr)-1);
}

File: csharp/mergeSort.cs
// マージソート
using System.IO;
using System;

class Program {
    static void Main() {
        int[] arr = {148, 651, 124, 638, 567, 435, 185, 413, 841, 35};
        Console.WriteLine(string.Join(", ", margeSort(arr)));
    }

    static int[] margeSort(int[] arr) {
        Func<int, int, int[]> m = null;
        m = (head, tail) => {
            if (head < tail) {
                int mid = (int)Math.Floor((double)(head + tail) / 2);
                m(head, mid);
                m(mid + 1, tail);
        
                int[] l = new int[mid+1-head];
                Array.Copy(arr, head, l, 0, mid+1-head);
                int i = 0;
                int j = mid + 1;
                int k = head;
                while (i < l.Length && j <= tail) {
                    if (l[i] <= arr[j]) {
                        arr[k++] = l[i++];
                    } else {
                        arr[k++] = arr[j++];
                    }
                }

                while (i < l.Length) {
                    arr[k++] = l[i++];
                }
            }
            return arr;
        };
        m(0, arr.Length-1);
        return arr;
    }
}
